### PR TITLE
zx: 7.2.3 -> 8.1.2

### PR DESCRIPTION
**Pull request by jlbribeiro**
## Description of changes

**Fixes** #309047.
**Changes maintainer** to @jlbribeiro, following [this comment](https://github.com/NixOS/nixpkgs/issues/309047#issuecomment-2095530702) by the current maintainer.

Given `zx` popularity and its usage as a shell scripting language, I assume it might be worth linking the 8.0.0 release changelog in the next NixOS release's notes.

~**edit:** As [we're currently in feature freeze](https://github.com/NixOS/nixpkgs/pull/309121#issuecomment-2097010035), I'm turning this PR into a draft, and will be updating the next release notes only when that document is created.~ :heavy_check_mark: 

**Diff**: https://github.com/google/zx/compare/7.2.3...8.1.2

**Changelog**:
- https://github.com/google/zx/releases/tag/8.0.0
- https://github.com/google/zx/releases/tag/8.0.1
- https://github.com/google/zx/releases/tag/8.0.2
- https://github.com/google/zx/releases/tag/8.1.0
- https://github.com/google/zx/releases/tag/8.1.1
- https://github.com/google/zx/releases/tag/8.1.2

**Maintainer**: @hlolli (currently), to become @jlbribeiro after merge

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/tools/system/zx/default.nix [hunk 1]
```diff
@@ -1,20 +1,24 @@
...
     repo = "zx";
     rev = version;
-    hash = "sha256-YMfecNazmL8J+f80FdIRvr2upQ7VgXSkQQnm8z0Swhw=";
+    hash = "sha256-ubf9D8teROF+Hn0EVlpJI4U8QCdhs1fZSDfzF9zrppE=";
   };
 
-  npmDepsHash = "sha256-ywNd2LGjM35ecW4dnj0oNwdSX2CRy8i9OGKPIdI0UEQ=";
+  npmDepsHash = "sha256-nE+Fo6uQnFsAJmqW/gDS/kr2llhxMgTVHZUomK5CS/s=";
+
+  ESBUILD_BINARY_PATH = "${lib.getExe esbuild}";
```
Comment (getchoo) [inline pkgs/tools/system/zx/default.nix:21]:
```suggestion
  ESBUILD_BINARY_PATH = lib.getExe esbuild;
```

**Review by getchoo — CHANGES_REQUESTED** [review]

File: nixos/doc/manual/release-notes/rl-2405.section.md [hunk 2]
```diff
@@ -238,6 +238,13 @@ The pre-existing [services.ankisyncd](#opt-services.ankisyncd.enable) has been m
 
 - `unrar` was updated to v7. See [changelog](https://www.rarlab.com/unrar7notes.htm) for more information.
 
+- `zx` was updated to v8, which introduces several breaking changes, namely:
+  `$.verbose` now defaults to `false` (but errors are still printed to `stderr`);
+  `ssh` API was dropped in favor of using the [webpod](https://github.com/webpod/webpod) package;
+  `powershell` is not automatically setup on Windows by default anymore;
```
Comment (kirillrdy) [inline nixos/doc/manual/release-notes/rl-2405.section.md:244]: is this relevant, since nixpkgs doesn't support windows ?

Comment (jlbribeiro) [reply nixos/doc/manual/release-notes/rl-2405.section.md:244]: @kirillrdy not gonna lie, I actually laughed out loud when I read your comment; I rewrote the changelog and didn't even give it a second thought. You're obviously right.

On a separate note: should I just link the changelog and avoid summarizing the changes themselves entirely? I saw a mixed of both approaches in the release notes.

**Review by hlolli — APPROVED** [review]

**Comment by kirillrdy** [comment]
@jlbribeiro there is currently breaking change restriction to all branches
 
as per definition of breaking change https://github.com/NixOS/nixpkgs/issues/303285

> Breaking change
> 
>     A change that is likely to break downstream consumers
> 


I know it's a leaf package, but unless someone says otherwise, this will have to wait for after 24.05 release

**Comment by jlbribeiro** [comment]
@kirillrdy That makes total sense, will have to wait then. This made me re-read the schedule and fully grasp the release cycle.
If I fully understood, I'll have to move the release notes' update to the future `24.??` (not `24.05`) release notes, correct?
(first time I noticed there's a `release-YY.mm` and `nixos-YY.mm`, was only aware of the latter)

(and again, thank you for the review + direction steering)

**Review by getchoo — APPROVED** [review]
renewing my approval, but as mentioned above we should wait until 24.05

**Comment by jlbribeiro** [comment]
> renewing my approval, but as mentioned above we should wait until 24.05

@getchoo Thank you.

Can you confirm, however, that this should **not be merged as-is** after 24.05 is released, given it would modify the wrong release notes? (I'm validating whether I understood the release cycle "rules")
Putting in in another way: did I understand correctly this package will not and must not reach the `nixos-24.05`, not even via a backport (as it would introduce breaking changes)? Just ensuring this doesn't get merged by mistake, given it's already approved. (couldn't find an appropriate label for this situation)

**Comment by kirillrdy** [comment]
> will not and must not reach the nixos-24.05, not even via a backport (as it would introduce breaking changes)

correct

>  Just ensuring this doesn't get merged by mistake

set PR to draft

**Comment by jlbribeiro** [comment]
@getchoo @kirillrdy kindly request your review (again) as a green light for getting this merged post feature freeze.

**Pull request merged**